src/Symfony/Component/Routing/Matcher/Dumper/PhpMatcherDumper.php

```diff
@@ -263,26 +266,57 @@ private function compileRoute(Route $route, $name, $supportsRedirections, $paren
 EOF;
 
         $gotoname = 'not_'.preg_replace('/[^A-Za-z0-9_]/', '', $name);
+
         if ($methods) {
             if (1 === count($methods)) {
-                $code .= <<<EOF
-            if (\$this->context->getMethod() != '$methods[0]') {
+                if ($methods[0] === 'HEAD') {
+                    $code .= <<<EOF
+            if ('HEAD' !== \$requestMethod) {
+                \$allow[] = 'HEAD';
+                goto $gotoname;
+            }
+
+
+EOF;
+                } else {
+                    $code .= <<<EOF
+            if ('$methods[0]' !== \$isLikeGetMethod) {
                 \$allow[] = '$methods[0]';
                 goto $gotoname;
             }
 
 
 EOF;
+                }
             } else {
-                $methods = implode("', '", $methods);
-                $code .= <<<EOF
-            if (!in_array(\$this->context->getMethod(), array('$methods'))) {
+                $methodVariable = 'requestMethod';
+
+                if (in_array('GET', $methods)) {
+                    // Since we treat HEAD requests like GET requests we don't need to match it.
+                    $methodVariable = 'isLikeGetMethod';
+                    $methods = array_filter($methods, function ($method) { return 'HEAD' !== $method; });
+                }
+
+                if (1 === count($methods)) {
+                    $code .= <<<EOF
+            if ('$methods[0]' !== \$$methodVariable) {
+                \$allow[] = '$methods[0]';
+                goto $gotoname;
```

Comment:
Was the use of the goto ever discussed?
It seems like here we only need to reverse the conditions to get ride of it.